src/openforms/forms/fd_translations_converter.py

```diff
@@ -0,0 +1,176 @@
+"""
+Implements the translation converter for form definitions.
+
+Github issue: https://github.com/open-formulieren/open-forms/issues/2958
+
+Before, translations for formio components were saved in the JSONField
+``FormDefinition.component_translations``, which is a key-value mapping of language
+code to translations mapping, which itself is a key-value mapping of source string to
+translation string. This had the drawback that if different components happen to use the
+same literal, they also shared the same translation.
+
+The utilities here help in converting these translations to component-local
+translations, where the translations are stored on each component inside the formio
+schema. The exact set of supported fields for translation varies per component, the
+reference for this is the typescript type definitions:
+https://open-formulieren.github.io/types/modules/index.html
+"""
+import logging
+from typing import TypedDict
+
+from glom import assign
+
+from openforms.formio.service import iter_components
+from openforms.formio.typing import Component
+
+logger = logging.getLogger(__name__)
+
+
+class Option(TypedDict):
+    value: str
+    label: str
+
+
+def _set_translations(
+    obj: Component | Option,
+    translations_from_store: dict[str, str],
+    locale: str,
+    fields: list[str],
+) -> None:
+    translations = {
+        key: translation
+        for key in fields
+        if (literal := obj.get(key))
+        and (translation := translations_from_store.get(literal))
+    }
```

Comment:
code is read more often than it is written, I find it hard to know what the resulting `translations` dict is. But fine to keep it as is

Reply:
yeah it's a side-effect of bad data structures that we're resolving :(

But it's basically making sure that the translation is grabbed from the `[literal]: [translation]` mapping if the literal is not empty-ish, and assigns it to the field/property where the literal was found (`label`, `description` etc...).

The frontend code for this is horrible :sweat_smile: